File: c/medium/minimum_swaps_to_arrange_a_binary_grid.c
#include <stdio.h>
#include <stdlib.h>

int minSwaps(int** grid, int n) {
    int* trailingZeros = (int*)malloc(sizeof(int) * n);

    // Count trailing zeros per row
    for (int i = 0; i < n; i++) {
        int count = 0;
        for (int j = n - 1; j >= 0; j--) {
            if (grid[i][j] == 0)
                count++;
            else
                break;
        }
        trailingZeros[i] = count;
    }

    int swapCount = 0;
    int currentSize = n;

    for (int targetRow = 0; targetRow < n; targetRow++) {
        int minRequiredZeros = n - 1 - targetRow;
        int candidateRow = -1;

        // find first valid row
        for (int i = 0; i < currentSize; i++) {
            if (trailingZeros[i] >= minRequiredZeros) {
                candidateRow = i;
                break;
            }
        }

        if (candidateRow == -1) {
            free(trailingZeros);
            return -1;
        }

        swapCount += candidateRow;

        // remove element (shift left)
        for (int i = candidateRow; i < currentSize - 1; i++) {
            trailingZeros[i] = trailingZeros[i + 1];
        }

        currentSize--;
    }

    free(trailingZeros);
    return swapCount;
}

File: rust/medium/minimum_swaps_to_arrange_a_binary_grid.rs
pub fn min_swaps(grid: Vec<Vec<i32>>) -> i32 {
    let n = grid.len();

    // Count trailing zeros
    let mut trailing_zeros: Vec<usize> = grid
        .iter()
        .map(|row| {
            row.iter()
                .rev()
                .take_while(|&&x| x == 0)
                .count()
        })
        .collect();

    let mut swap_count = 0;

    for target_row in 0..n {
        let min_required_zeros = n - 1 - target_row;

        let candidate_row = trailing_zeros
            .iter()
            .position(|&z| z >= min_required_zeros);

        match candidate_row {
            Some(idx) => {
                swap_count += idx as i32;
                trailing_zeros.remove(idx);
            }
            None => return -1,
        }
    }

    swap_count
}